lib/kcerts/ssh.go

```diff
@@ -0,0 +1,116 @@
+package kcerts
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"fmt"
+	"github.com/mitchellh/go-homedir"
+	"golang.org/x/crypto/ssh"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+)
+
+const CAPrefix = "@cert-authority"
+const SSHDir = ".ssh"
+const KnownHosts = "known_hosts"
+
+// AddSSHCAToClient adds a public key to the $HOME/.ssh/known_hosts in the ssh-cert x509.1 format.
+// For each entry, it adds an additional line and does not concatenate
+func AddSSHCAToClient(publicKey ssh.PublicKey, hosts []string) error {
+	hDir, err := homedir.Dir()
+	if err != nil {
+		return err
+	}
+	hDir = hDir + "/"
+	if _, err := os.Stat(hDir + SSHDir); os.IsNotExist(err) {
+		return fmt.Errorf("ssh directory %s does not exist, please create it", hDir+SSHDir)
+	}
+	qualifiedKnownHosts := hDir + SSHDir + "/" + KnownHosts
+	if _, err := os.Stat(qualifiedKnownHosts); os.IsNotExist(err) {
+		return fmt.Errorf("ssh authorized hosts file %s does not exist, please create it", qualifiedKnownHosts)
+	}
+	caPublic := string(ssh.MarshalAuthorizedKey(publicKey))
+	existingKnownHostsContent, err := ioutil.ReadFile(qualifiedKnownHosts)
+	if err != nil {
+		return err
+	}
+	knownHostsFile, err := os.OpenFile(qualifiedKnownHosts, os.O_APPEND|os.O_WRONLY, 0644)
+	if err != nil {
+		return err
+	}
+	defer knownHostsFile.Close()
+	for _, dns := range hosts {
+		publicFormat := fmt.Sprintf("%s %s %s", CAPrefix, dns, caPublic)
+		if !strings.Contains(string(existingKnownHostsContent), publicFormat) {
+			_, err = knownHostsFile.WriteString(publicFormat)
+			if err != nil {
+				fmt.Println("error is not nil", err.Error())
+				return err
```

Comment:
I'm a big fan of the [new error handling introduced in go1.13](https://blog.golang.org/go1.13-errors). 

Let's say the disk is full, and this error is printed in the logs. You go debug it, all you'll see is "no space on device".

My recommendation is to return:
`fmt.Errorf("could not add key %s to known_hosts file: %w", err)`

(note the '%w', and err passed as is) 
This would generate a very readable and much more informative error message (now I know the implication of no space on disk), while still allowing easy programmatic access by using the `Unwrap` or `errors.Is`, `errors.As` API.

Reply:
At some point I'm going to have to bump the go version.
Been using the go's errors.Wrap instead of the fmt for a while, super nice. 
https://pkg.go.dev/github.com/pkg/errors#Wrap